festim/source.py

```diff
@@ -0,0 +1,198 @@
+import festim as F
+import ufl
+from dolfinx import fem
+import numpy as np
+
+
+class Source:
+    """
+    Source class
+
+    Args:
+        volume (festim.VolumeSubdomain1D): the volume subdomain where the source
+            is applied
+        value (float, int, fem.Constant or callable): the value of the soure
+        species (str): the name of the species to which the source is applied
+
+    Attributes:
+        volume (festim.VolumeSubdomain1D): the volume subdomain where the source
+            is applied
+        value (float, int, fem.Constant or callable): the value of the soure
+        species (str): the name of the species to which the source is applied
+        value_fenics (fem.Function or fem.Constant): the value of the source in
+            fenics format
+        source_expr (fem.Expression): the expression of the source term that is
+            used to update the value_fenics
+        time_dependent (bool): True if the value of the source is time dependent
+        temperature_dependent (bool): True if the value of the source is temperature
+            dependent
+
+    Usage:
+        >>> from festim import Source
+        >>> Source(volume=my_vol, value=1, species="H")
+        >>> Source(volume=my_vol, value=lambda x: 1 + x[0], species="H")
+        >>> Source(volume=my_vol, value=lambda t: 1 + t, species="H")
+        >>> Source(volume=my_vol, value=lambda T: 1 + T, species="H")
+        >>> Source(volume=my_vol, value=lambda x, t: 1 + x[0] + t, species="H")
+    """
+
+    def __init__(self, value, volume, species):
+        self.value = value
+        self.volume = volume
+        self.species = species
+
+        self.value_fenics = None
+        self.source_expr = None
+
+    @property
+    def volume(self):
+        return self._volume
+
+    @volume.setter
+    def volume(self, value):
+        # check that volume is festim.VolumeSubdomain1D or list of festim.VolumeSubdomain1D
+        if not isinstance(value, (F.VolumeSubdomain1D, int)) and not isinstance(
+            value, list
+        ):
+            raise TypeError(
+                "volume must be of type festim.VolumeSubdomain1D or int or a list of festim.VolumeSubdomain1D or int"
+            )
+        # check that all elements of list are festim.VolumeSubdomain1D
+        if isinstance(value, list):
+            for element in value:
+                if not isinstance(element, (F.VolumeSubdomain1D, int)):
+                    raise TypeError(
+                        "volume must be of type festim.VolumeSubdomain1D or int or a list of festim.VolumeSubdomain1D or int"
+                    )
+        # if volume is festim.VolumeSubdomain1D, convert to list
+        if not isinstance(value, list):
+            value = [value]
```

Comment:
I'm not sure if we should do this in the setter as it modifies the user input quite a lot... Maybe we can do this in HTransportProblem instead